src/names.c

```diff
@@ -33,20 +33,25 @@ SEXP vec_as_names(SEXP names, const struct name_repair_opts* opts) {
   case name_repair_minimal: return vctrs_as_minimal_names(names);
   case name_repair_unique: return vec_as_unique_names(names, opts->quiet);
   case name_repair_universal: return vec_as_universal_names(names, opts->quiet);
-  case name_repair_check_unique: return vec_validate_unique_names(names);
+  case name_repair_check_unique: return vec_validate_unique_names(names, opts->arg);
   case name_repair_custom: return vec_as_custom_names(names, opts);
   }
   never_reached("vec_as_names");
 }
 
 // [[ register() ]]
-SEXP vctrs_as_names(SEXP names, SEXP repair, SEXP quiet) {
+SEXP vctrs_as_names(SEXP names, SEXP repair, SEXP repair_arg, SEXP quiet) {
   if (!r_is_bool(quiet)) {
     Rf_errorcall(R_NilValue, "`quiet` must a boolean value.");
   }
   bool quiet_ = LOGICAL(quiet)[0];
 
-  struct name_repair_opts repair_opts = new_name_repair_opts(repair, quiet_);
+  if (!r_is_string(repair_arg)) {
+    Rf_errorcall(R_NilValue, "`repair_arg` must a string.");
+  }
```

Comment:
The validation should be left to `vec_as_arg()`. In the future this could accept quoted symbols/calls.